core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java

```diff
@@ -0,0 +1,218 @@
+  }
+
+  private static Record fromManifestEntry(
+      ManifestEntry<?> entry, Table table, Schema recordSchema) {
+    GenericRecord record = GenericRecord.create(recordSchema);
+    Types.StructType partitionType =
+        recordSchema.findField(Column.PARTITION.name()).type().asStructType();
+    Record partitionData = coercedPartitionData(entry.file(), table.specs(), partitionType);
+    record.set(Column.PARTITION.ordinal(), partitionData);
+    record.set(Column.SPEC_ID.ordinal(), entry.file().specId());
+
+    Snapshot snapshot = table.snapshot(entry.snapshotId());
+    if (snapshot != null) {
+      record.set(Column.LAST_UPDATED_SNAPSHOT_ID.ordinal(), snapshot.snapshotId());
+      record.set(Column.LAST_UPDATED_AT.ordinal(), snapshot.timestampMillis());
+    }
+
+    switch (entry.file().content()) {
+      case DATA:
+        record.set(Column.DATA_RECORD_COUNT.ordinal(), entry.file().recordCount());
+        record.set(Column.DATA_FILE_COUNT.ordinal(), 1);
+        record.set(Column.TOTAL_DATA_FILE_SIZE_IN_BYTES.ordinal(), entry.file().fileSizeInBytes());
+        break;
+      case POSITION_DELETES:
+        record.set(Column.POSITION_DELETE_RECORD_COUNT.ordinal(), entry.file().recordCount());
+        record.set(Column.POSITION_DELETE_FILE_COUNT.ordinal(), 1);
+        record.set(Column.DATA_RECORD_COUNT.ordinal(), 0L);
+        record.set(Column.DATA_FILE_COUNT.ordinal(), 0);
+        record.set(Column.TOTAL_DATA_FILE_SIZE_IN_BYTES.ordinal(), 0L);
+        break;
+      case EQUALITY_DELETES:
+        record.set(Column.EQUALITY_DELETE_RECORD_COUNT.ordinal(), entry.file().recordCount());
+        record.set(Column.EQUALITY_DELETE_FILE_COUNT.ordinal(), 1);
+        record.set(Column.DATA_RECORD_COUNT.ordinal(), 0L);
+        record.set(Column.DATA_FILE_COUNT.ordinal(), 0);
+        record.set(Column.TOTAL_DATA_FILE_SIZE_IN_BYTES.ordinal(), 0L);
+        break;
+      default:
+        throw new UnsupportedOperationException(
+            "Unsupported file content type: " + entry.file().content());
+    }
+
+    // Note: Not computing the `TOTAL_RECORD_COUNT` for now as it needs scanning the data.
+    return record;
+  }
+
+  private static Record coercedPartitionData(
+      ContentFile<?> file, Map<Integer, PartitionSpec> specs, Types.StructType partitionType) {
+    // keep the partition data as per the unified spec by coercing
+    StructLike partition =
+        PartitionUtil.coercePartition(partitionType, specs.get(file.specId()), file.partition());
+    GenericRecord genericRecord = GenericRecord.create(partitionType);
+    for (int index = 0; index < partitionType.fields().size(); index++) {
+      genericRecord.set(
+          index,
+          partition.get(index, partitionType.fields().get(index).type().typeId().javaClass()));
+    }
+
+    return genericRecord;
+  }
+
+  private static void checkAndIncrementLong(Record toUpdate, Record fromRecord, Column column) {
+    if (fromRecord.get(column.ordinal()) != null) {
```

Comment:
Personal prefrence from me here but I find it clearer (albeit longer) to do
```java
if (from !=null && to !=null) {
   to = from + to
} else if (from != null) {
   to = from;
}
```

Just to decrease the nesting